packages/malloy/src/lang/ast/types/malloy-element.ts

```diff
@@ -291,13 +306,19 @@ export class ModelEntryReference extends MalloyElement {
 }
 
 export interface DocStatement extends MalloyElement {
-  execute(doc: Document): ModelDataRequest;
+  execute(doc: Document): void;
 }
 
 export function isDocStatement(e: MalloyElement): e is DocStatement {
   return (e as DocStatement).execute !== undefined;
 }
 
+export function isDocStatementOrDocStatementList(
```

Comment:
> First of all, a doc statement which defines a lot of things, that is just a DefinitionList. There should be a way for a thing to be a doc statement and a definition list. If you can't do that with inheritance, then "am i a doc statement" is an interface.

So the two subclasses of `DocStatementList` are `DefineSourceList` and `DefineQueryList`, which behave slightly differently from the other `DefinitionList`s in that they contain a list of top level definition statements which may need to be executed before subsequent ones can be compiled (i.e. they have the `executeList` method). Previously, `DocStatementList` was called `RunList`, and it itself was a `DocStatement`, which was a little confusing, because it felt like a "document statement" should be the unit of sequential execution (but in reality, the individual statements inside a `RunList`/`DocStatementList` need to be run sequentially). 

> Second, now that you have invented this piece of brilliance, all the "execute cursor" stuff should go away. Document.compile should just return needs from children until there are no needs, and then just call execute on each statement.

I don't think I can do that, since SQL expressions may still depend on prior statements being executed (e.g. they might need a source to have been defined already).

Maybe we can hop on a call today to chat about these things.